### PR TITLE
Marginalize out discrete latent sites in HMC via enumeration

### LINKED ISSUE #1128: Automatically enumerate discrete variables in HMC
## Why?

Lots of models have discrete latent variables, e.g. mixture models. See e.g. https://forum.pyro.ai/t/nuts-with-discrete-latent-site-very-slow/204

## How?

Following `TraceEnum_ELBO` it should be possible to marginalize out any finite discrete latent variables:
1. wrap the model in `@config_enumerate`
2. use `EnumerateMessenger` to generate a bag of enumerated model trace (multiple traces come from sequential enumeration; broadcasted tensors come from parallel enumeration; both are reasonable)
3. block the enumerated sites from HMC and instead using their `site['log_prob'].exp()` to multiply downstream `site['log_prob']`.

### PR DESCRIPTION
**Pull request by neerajprad**
Resolves #1128 via parallel enumeration.

 - If a model has discrete latents, they will be enumerated over in parallel. In such cases, we need to supply `max_iarange_nesting` as arg to HMC/NUTS constructor to reserve dims for parallel enumeration. Also in these cases, we do not allow for stray batch dims, i.e. batch dimensions need to be wrapped inside an `iarange`.
 - Tested on gmm and a latent bernoulli model. Also, added some cheap tests in `mcmc/test_valid_models.py` to assert error/ok behavior.

### CODE REVIEW

File: pyro/infer/mcmc/hmc.py [hunk 1]
```diff
@@ -93,6 +108,35 @@ def _get_trace(self, z):
...
+                    if site["type"] == "sample"}
+
+        log_probs = defaultdict(float)
+        for name, site in model_trace.nodes.items():
+            if site["type"] == "sample":
+                log_probs[ordering[name]] = log_probs[ordering[name]] + site["log_prob"]
+
+        log_prob_sum = 0.
+        for log_prob in log_probs.values():
+            while log_prob.dim() > self.max_iarange_nesting:
```
Comment (fritzo) [inline pyro/infer/mcmc/hmc.py]: nit: it would.be cheaper to reshape, something like:
```py
enum_dim = log_prob.dim() - self.max_iarange_nesting
if enum_dim > 0:
    log_prob = log_sum_exp(log_prob.reshape(-1, *log_prob.shape[enum_dim:]), dim=0)
```

Comment (neerajprad) [reply pyro/infer/mcmc/hmc.py]: +1; will make the change.

**Comment by fritzo** [comment]
Do I understand correctly that:
1. This PR implements parallel enumeration but not sequential enumeration.
2. The restriction to parallel enumeration is reasonable because HMC already makes the assumption that model structure is fixed, and the main use case of sequential enumeration is for models with dynamic structure.

File: pyro/infer/mcmc/hmc.py [hunk 2]
```diff
@@ -93,6 +108,36 @@ def _get_trace(self, z):
...
+        log_probs = defaultdict(float)
+        for name, site in model_trace.nodes.items():
+            if site["type"] == "sample":
+                log_probs[ordering[name]] = log_probs[ordering[name]] + site["log_prob"]
+
+        log_prob_sum = 0.
+        for log_prob in log_probs.values():
+            enum_dim = log_prob.dim() - self.max_iarange_nesting
+            if enum_dim > 0:
+                log_prob = log_sum_exp(log_prob.reshape(-1, *log_prob.shape[enum_dim:]), dim=0)
```
Comment (fritzo) [inline pyro/infer/mcmc/hmc.py]: I believe you'll need to do something fancier like `Dice.in_context()` here. At the moment, you're summing over only the local probabilities rather than all probabilities. This will fail for models with nested enumeration, e.g.
```py
@poutine.broadcast
def model(data):
    p = pyro.sample("p", dist.Uniform(0., 1.))
    y = pyro.sample("y", dist.Bernoulli(p))
    q = 0.5 + 0.25 * y
    with pyro.iarange("data", len(data)):
        z = pyro.sample("z", dist.Bernoulli(q))
        mean = 2 * z - 1
        pyro.sample("obs", dist.Normal(mean, 1.), obs=data)
```
In this case I believe you'll currently compute
```py
p(p) = p_unif(p)
     + sum(p_bern(y | p) for y in [0,1])
     + prod(sum(p_bern(zi | q(y)) for zi in [0,1] for y in [0,1])
            for i in range(len(data)))
```
which is not equal to the intended
```py
p(p) = p_unif(p)
     + sum(p_bern(y | p) * prod(sum(p_bern(zi | q(y)) for zi in [0,1])
                                for i in range(len(data)))
           for y in [0,1])
```
Could we add something like this as a unit test against a vetted `log_prob` computation?

**Comment by fritzo** [comment]
Maybe it's worth moving `_compute_dice_elbo()` from traceenum_elbo.py into the `Dice` class as a `Dice.compute_elbo()` method, and then implementing what you want as a `Dice.compute_log_prob()` method? It seems like their code bodies are quite similar, but maybe their assumptions are sufficiently different for them to more naturally lie closer to their callers?

**Comment by neerajprad** [comment]
> EDIT I've attempted to do some of this in #1251, hopefully that helps a bit.

Thanks for the PR, @fritzo! Let me go through this in more detail, and reply to your comments.

**Comment by neerajprad** [comment]
> Do I understand correctly that:

> This PR implements parallel enumeration but not sequential enumeration. The restriction to parallel enumeration is reasonable because HMC already makes the assumption that model structure is fixed, and the main use case of sequential enumeration is for models with dynamic structure.

That's right. Sequential enumeration will greatly add to the already considerable computational overhead (since I think we'll need to run the integrator on each trace separately), making it not as useful practically.

File: pyro/infer/mcmc/hmc.py (hunk 2 above)
Comment (neerajprad) [reply pyro/infer/mcmc/hmc.py]: Thanks for this example, really helped me understand the issue with nested enumeration. I will add a test to validate the log prob computation.

**Comment by eb8680** [comment]
> Sequential enumeration will greatly add to the already considerable computational overhead (since I think we'll need to run the integrator on each trace separately), making it not as useful practically.

This could still be very useful for models whose log-probability computation graph after enumeration is static, since those can still be accelerated by the PyTorch JIT.  I don't think we need to handle the sequential case in this PR now.

**Comment by neerajprad** [comment]
Updated the PR with the following:
 - The log prob computation used `Dice.in_context` to get the weights from the enumerated sites.
 - Added a test that validates the log prob computation for nested enumeration (can be computed by hand). Also modified the latent bernoulli test to catch the case explained by @fritzo.

I am seeing that using `Dice` and computing `score_parts` has increased the runtime by 40-50%, so there might be some opportunity for optimizing specifically for HMC.

**Comment by neerajprad** [comment]
I could not use `Dice.compute_expectation` because the cost (probs of continuous sites) factors do not add up between different ordinals (but rather need to multiplied together). Similarly the dims within `max_iarange_nesting` need to be multiplied. If we use `log_probs` instead, that leads to something which isn't as general, and is basically what we have in `HMC._compute_trace_log_prob` currently.

File: pyro/infer/mcmc/hmc.py [hunk 3]
```diff
@@ -93,6 +109,41 @@ def _get_trace(self, z):
...
+                             "has discrete (enumerable) sites.")
+        model_trace.compute_log_prob()
+        model_trace.compute_score_parts()
+        ordering = {name: frozenset(site["cond_indep_stack"])
+                    for name, site in model_trace.nodes.items()
+                    if site["type"] == "sample"}
+
+        log_probs = defaultdict(float)
+        for name, site in model_trace.nodes.items():
+            if site["type"] == "sample" and not site["fn"].has_enumerate_support:
```
Comment (fritzo) [inline pyro/infer/mcmc/hmc.py]: I believe this incorrectly ignores discrete observe sites?

Comment (neerajprad) [reply pyro/infer/mcmc/hmc.py]: Thanks for catching, will change and a test for this.

File: pyro/infer/mcmc/util.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
+from collections import defaultdict
+
+from pyro.distributions.util import log_sum_exp
+from pyro.infer.util import is_validation_enabled
+from pyro.util import check_site_shape
+
+
+class EnumTraceProbEvaluator(object):
+    def __init__(self,
```
Comment (neerajprad) [inline pyro/infer/mcmc/util.py]: @frtizo - As discussed, I have removed the dependency on Dice, and moved the log prob computation to its own class to make it easy to test (test cases are added in `infer/mcmc/test_valid_models.py`). I think this should be self-explanatory (and very similar to the Dice logic), but let me know in case I may have overlooked some corner cases.

Comment (neerajprad) [reply pyro/infer/mcmc/util.py]: Also, this should resolve the bugs in the previous implementation. I realized that the logic of only looking at the enumerated dims when computing the upstream log probs was wrong.

File: pyro/infer/mcmc/util.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+                 model_trace,
+                 has_enumerable_sites=False,
+                 max_iarange_nesting=float("inf")):
+        """
```
Comment (fritzo) [inline pyro/infer/mcmc/util.py]: nit: could you move this comment to a class-level comment rather than an `__init__` comment?

File: pyro/infer/mcmc/util.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        :param bool has_enumerable_sites: whether the trace contains any
+            discrete enumerable sites.
+        :param int max_iarange_nesting: Optional bound on max number of nested
+            :func:`pyro.iarange` contexts.
+        """
+        self.model_trace = model_trace
+        self.has_enumerable_sites = has_enumerable_sites
+        self.max_iarange_nesting = max_iarange_nesting
+        self.log_probs = defaultdict(list)
+        self._log_factors_cache = {}
```
Comment (fritzo) [inline pyro/infer/mcmc/util.py]: nit: It's cheaper and probably simpler to use a cache of pairs rather than a pair of caches.

File: tests/infer/mcmc/test_valid_models.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+    @poutine.broadcast
+    def model(data):
+        p = pyro.sample("p", dist.Uniform(0., 1.))
+        y = pyro.sample("y", dist.Bernoulli(p))
+        mean = 2 * y - 1
+        n = pyro.sample("n", dist.Normal(mean, 1.))
+        with pyro.iarange("data", len(data)):
+            pyro.sample("obs", dist.Bernoulli(sigmoid(n)), obs=data)
+
+    conditioned_model = poutine.enum(
```
Comment (fritzo) [inline tests/infer/mcmc/test_valid_models.py]: nit: why not add these to decorators above?
```py
@poutine.enum(first_available_dim=1)
@config_enumerate(default="parallel")
@poutine.condition(data={"p": ..., "n": ...})
@poutine.broadcast
def model(data):
    ...
```

File: pyro/infer/mcmc/util.py (hunk 5 above)
Comment (neerajprad) [reply pyro/infer/mcmc/util.py]: My bad, will change.

File: pyro/infer/mcmc/util.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+            return self.model_trace.log_prob_sum()
+        if self.max_iarange_nesting == float("inf"):
+            raise ValueError("Finite value required for `max_iarange_nesting` when model "
+                             "has discrete (enumerable) sites.")
+        self._compute_log_prob_terms()
+
+        # Sum up terms from predecessor, and gather leaf nodes.
+        leaves_log_probs = {}
+        for target_ordinal in sorted(self.log_probs.keys()):
+            leaves_log_probs[target_ordinal] = self.log_probs[target_ordinal]
```
Comment (fritzo) [inline pyro/infer/mcmc/util.py]: nit: this is a bit confusing. Consider simplifying to
```py
for target_ordinal, log_prob in self.log_probs.items():
    if any(target_ordinal < other for other in self.log_probs):
        continue  # not a leaf
    predecessors, log_factors = self._get_predecessors_log_factors(target_ordinal)
    leaves_log_probs[target_ordinal] = sum(log_prob + log_factors)
```

File: pyro/infer/mcmc/util.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+    def _get_predecessors_log_factors(self, target_ordinal):
+        """
+        Returns the list of predecessors for `target_ordinal `and
+        their log_prob factors.
+        """
+        if target_ordinal in self._predecessors_cache:
+            return self._predecessors_cache[target_ordinal]
+        log_factors = []
+        predecessors = set()
```
Comment (fritzo) [inline pyro/infer/mcmc/util.py]: Am I correct that `predecessors` is no longer used, so we can remove this data structure?

File: pyro/infer/mcmc/util.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+            predecessors, log_factors = self._get_predecessors_log_factors(target_ordinal)
+            leaves_log_probs[target_ordinal] = sum(leaves_log_probs[target_ordinal] + log_factors)
+
+        # Reduce the log prob terms for each leaf node:
+        # - taking log_sum_exp of factors in enum dims (i.e.
+        # adding up the probability terms).
+        # - summing up the dims within `max_iarange_nesting`.
+        # (i.e. multiplying probs within independent batches).
+        log_prob_sum = 0.
+        for ordinal in leaves_log_probs:
```
Comment (fritzo) [inline pyro/infer/mcmc/util.py]: nit: simplfy using `.items()`
```py
for ordinal, log_prob in leaves_log_probs.items():
    ...
```

File: pyro/infer/mcmc/util.py (hunk 9 above)
Comment (neerajprad) [reply pyro/infer/mcmc/util.py]: Fixed.

File: pyro/infer/mcmc/util.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        if not self.has_enumerable_sites:
+            return self.model_trace.log_prob_sum()
+        if self.max_iarange_nesting == float("inf"):
+            raise ValueError("Finite value required for `max_iarange_nesting` when model "
+                             "has discrete (enumerable) sites.")
+        self._compute_log_prob_terms()
+
+        # Sum up terms from predecessor, and gather leaf nodes.
+        leaves_log_probs = {}
+        for target_ordinal in sorted(self.log_probs.keys()):
```
Comment (fritzo) [inline pyro/infer/mcmc/util.py]: nit: simplify using `.items()`
```py
for target_ordinal, log_prob in self.log_probs.items():
    if ...
        continue:
    log_factors = self._get_log_factors(target_ordinal)
    leaves_log_probs[target_ordinal] = log_prob + sum(log_factors)
```

**Comment by fritzo** [comment]
> we just need to multiply the terms per independence context

I guess you could aggregate up the tree, starting with leaves and moving outward/upward. As you leave each independence context you `prod(sum(-))`? Another approach that I've taken in sequential enumeration is to explicitly scale for multiple counting by accumulating a `denominators` dict keyed on `ordinal`. Whatever works :smile:

**Comment by neerajprad** [comment]
> I guess you could aggregate up the tree, starting with leaves and moving outward/upward. As you leave each independence context you prod(sum(-))?

This is the approach I have taken now. It cleanly solves a bunch of issues we were getting earlier, but this does not work for general DAGs (i.e., indep contexts with multiple parents).

File: pyro/infer/mcmc/util.py [hunk 12]
```diff
@@ -0,0 +1,137 @@
+import torch
+from collections import defaultdict
+
+from pyro.distributions.util import log_sum_exp
+from pyro.infer.util import is_validation_enabled
+from pyro.poutine.indep_messenger import CondIndepStackFrame
+from pyro.util import check_site_shape
+
+
+class EnumTraceProbEvaluator(object):
```
Comment (neerajprad) [inline pyro/infer/mcmc/util.py]: @fritzo - This is ready for another round of review. 😄The important change is confined to this class. In HMC we are calling the constructor once to populate the children nodes and compute which dims should be evaluated/enumerated out. When we call `log_prob` subsequently, we reuse this computed metadata to aggregate the log prob terms from the leaf nodes upwards.

File: pyro/infer/mcmc/hmc.py [hunk 13]
```diff
@@ -93,6 +108,15 @@ def _get_trace(self, z):
         trace_poutine(*self._args, **self._kwargs)
         return trace_poutine.trace
 
+    @staticmethod
+    def _iter_latent_nodes(trace):
+        for name, node in sorted(trace.iter_stochastic_nodes(), key=lambda x: x[0]):
```
Comment (fritzo) [inline pyro/infer/mcmc/hmc.py]: Why sort by name rather than the default order (by execution time)?

File: pyro/infer/mcmc/util.py [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+    :param int max_iarange_nesting: Optional bound on max number of nested
+        :func:`pyro.iarange` contexts.
+    """
+    def __init__(self,
+                 model_trace,
+                 has_enumerable_sites=False,
+                 max_iarange_nesting=float("inf")):
+        self.has_enumerable_sites = has_enumerable_sites
+        self.max_iarange_nesting = max_iarange_nesting
+        default_cond_stack = (CondIndepStackFrame(name="default", dim=0, size=0, counter=None),)
```
Comment (fritzo) [inline pyro/infer/mcmc/util.py]: Why not the empty tuple?
```py
self.root = frozenset((()))
```

File: pyro/infer/mcmc/hmc.py (hunk 13 above)
Comment (neerajprad) [reply pyro/infer/mcmc/hmc.py]: Yeah, any fixed order is fine. The advantage to using names is that the dict based data structures that we are using can be sorted anytime by key names (otherwise, we'll need to store the key order somewhere and then sort using that order).

File: pyro/infer/mcmc/util.py [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+        self.root = frozenset(default_cond_stack)
```
Comment (fritzo) [inline pyro/infer/mcmc/util.py]: Why not the empty `frozenset()`?

Comment (neerajprad) [reply pyro/infer/mcmc/util.py]: I think this can be removed now. I was initially doing something like `if frozenset() in frozenset([1])` (which evaluates to False), so this was to not have to put in special cases to deal with the root node. Will remove this now.

File: pyro/infer/mcmc/util.py [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+    def _compute_log_prob_terms(self, model_trace):
+        """
+        Computes the conditional probabilities for each of the sites
+        in the model trace, and stores the result in `self._log_probs`.
+        """
+        model_trace.compute_log_prob()
+        ordering = {}
+        for name, site in model_trace.nodes.items():
+            if site["type"] == "sample":
+                if len(site["cond_indep_stack"]) == 0:
```
Comment (fritzo) [inline pyro/infer/mcmc/util.py]: If you make the `root` an empty frozenset then these four lines simplify to
```py
ordering[name] = frozenset(site["cond_indep_stack"])
```

Comment (neerajprad) [reply pyro/infer/mcmc/util.py]: Changed, having a special default key isn't needed anymore.

Comment (fritzo) [reply pyro/infer/mcmc/util.py]: or simply
```py
ordering = {name: frozenset(site["cond_indep_stack"])
            for name, site in model_trace.nodes.items()
            if site["type"] == "sample"}
```

File: pyro/infer/mcmc/util.py [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+
+        # Collect log prob terms per independence context.
+        log_probs = defaultdict(list)
+        for name, site in model_trace.nodes.items():
+            if site["type"] == "sample":
+                if is_validation_enabled():
+                    check_site_shape(site, self.max_iarange_nesting)
+                log_probs[ordering[name]].append(site["log_prob"])
+
+        for ordinal in log_probs:
```
Comment (fritzo) [inline pyro/infer/mcmc/util.py]: nit: If you use `.items()` you can avoid half of the hashing of frozensets.
```py
for ordinal, terms in log_probs.items():
    self._log_probs[ordinal] = sum(terms)
```
Also, cool trick to avoid `0.0 + x` tensor ops!

File: pyro/infer/mcmc/util.py [hunk 18]
```diff
@@ -0,0 +1,137 @@
+import torch
+from collections import defaultdict
+
+from pyro.distributions.util import log_sum_exp
```
Comment (fritzo) [inline pyro/infer/mcmc/util.py]: Could we try to use the `logsumexp` form internally, since that's what PyTorch 0.4.1 calls it?

Comment (neerajprad) [reply pyro/infer/mcmc/util.py]: Oh..sorry, I forgot to update. Will do so.

File: pyro/infer/mcmc/util.py [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+        for name, site in model_trace.nodes.items():
+            if site["type"] == "sample":
+                if is_validation_enabled():
+                    check_site_shape(site, self.max_iarange_nesting)
+                log_probs[ordering[name]].append(site["log_prob"])
+
+        for ordinal in log_probs:
+            self._log_probs[ordinal] = sum(log_probs[ordinal])
+
+    def _reduce(self, ordinal, agg_log_prob=torch.tensor(0.)):
```
Comment (fritzo) [inline pyro/infer/mcmc/util.py]: It's dangerous to create a `torch.tensor` at library load time, since it will not pick up `torch.set_default_tensor_type()` invocations. Instead I'd default to `None` and to an `if agg_log_prob is not None:` check.

Comment (neerajprad) [reply pyro/infer/mcmc/util.py]: Good catch! Will change.

File: pyro/infer/mcmc/util.py [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+            log_prob = log_sum_exp(log_prob, dim=enum_dim, keepdim=True)
+        for marginal_dim in self._iarange_dims[ordinal]:
+            log_prob = log_prob.sum(dim=marginal_dim, keepdim=True)
+        return log_prob
+
+    def _aggregate_log_probs(self, ordinal):
+        """
+        Aggregate the `log_prob` terms using depth first search.
+        """
+        if self._children[ordinal] is None:
```
Comment (fritzo) [inline pyro/infer/mcmc/util.py]: Hmm `self._children` is a `defaultdict(list)`. Do you instead mean to
```py
if not self._children[ordinal]:
    return self._reduce(ordinal)
```

File: pyro/infer/mcmc/util.py [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+            return self._reduce(ordinal)
+        agg_log_prob = torch.tensor(0.)
```
Comment (fritzo) [inline pyro/infer/mcmc/util.py]: You could avoid `torch.tensor` and save a tensor op by instead using python's `sum`:
```py
agg_log_prob = sum(map(self._aggregate_log_probs, self._children[ordinal]))
```

File: pyro/infer/mcmc/util.py (hunk 19 above)
Comment (neerajprad) [reply pyro/infer/mcmc/util.py]: I think it will be safe here to just use a python scalar. If a trace has no enumerable sites, we won't be going through this code path anyways, and if it has this 0. will get added as expected on the correct device.

File: pyro/infer/mcmc/util.py [hunk 22]
```diff
@@ -0,0 +1,137 @@
...
+        """
+        Returns the log pdf of `model_trace` by appropriately handling
+        enumerated log prob factors.
+
+        :return: log pdf of the trace.
+        """
+        if not self.has_enumerable_sites:
+            return model_trace.log_prob_sum()
+        self._compute_log_prob_terms(model_trace)
+        return self._aggregate_log_probs(self.root).sum()
```
Comment (fritzo) [inline pyro/infer/mcmc/util.py]: Do I understand correctly that:
1. all enumeration dims are to the left of `-max_iarange_nesting`, and they will have already been summed out by `self._aggregate_log_probs`
2. the result of `self._aggregate_log_probs(self.root)` will have non-reduced dims in exactly the global iaranges, i.e. the iaranges inside of which all sample statements are called, e.g. parallel particle iaranges?

File: pyro/infer/mcmc/util.py [hunk 23]
```diff
@@ -0,0 +1,137 @@
...
+        """
+        For each ordinal, populate the `iarange` and `enum` dims to be
+        evaluated or enumerated out.
+        """
+        log_prob = self._log_probs[ordinal]
+        iarange_dims = sorted([frame.dim for frame in ordinal - parent_ordinal])
+        enum_dims = set((i for i in range(-log_prob.dim(), -self.max_iarange_nesting)
+                         if log_prob.shape[i] > 1))
+        self._iarange_dims[ordinal] = iarange_dims
+        self._enum_dims[ordinal] = sorted(list(enum_dims - parent_enum_dims))
```
Comment (fritzo) [inline pyro/infer/mcmc/util.py]: nit: no need for `list( )`

File: pyro/infer/mcmc/util.py [hunk 24]
```diff
@@ -0,0 +1,137 @@
...
+    def _parse_model_structure(self, model_trace):
+        if not self.has_enumerable_sites:
+            return
+        if self.max_iarange_nesting == float("inf"):
+            raise ValueError("Finite value required for `max_iarange_nesting` when model "
+                             "has discrete (enumerable) sites.")
+        self._compute_log_prob_terms(model_trace)
+        # 1. Infer model structure - compute parent-child relationship.
+        sorted_ordinals = sorted(self._log_probs.keys())
+        for i in range(len(sorted_ordinals)):
```
Comment (fritzo) [inline pyro/infer/mcmc/util.py]: nit: consider simplifying to
```py
for i, child_node in enumerate(sorted_ordinals):
```

**Comment by neerajprad** [comment]
I think your comment got collapsed:
> Do I understand correctly that: all enumeration dims are to the left of -max_iarange_nesting, and they will have already been summed out by self._aggregate_log_probs

That's correct! 

> the result of self._aggregate_log_probs(self.root) will have non-reduced dims in exactly the global iaranges, i.e. the iaranges inside of which all sample statements are called, e.g. parallel particle iaranges?

They should be reduced dims, except that they will all be singleton, i.e. like `[[[1.3161]]]`. We don't currently have any global iaranges since we are not running parallel chains.

File: pyro/infer/mcmc/util.py (hunk 24 above)
Comment (neerajprad) [reply pyro/infer/mcmc/util.py]: done.

File: pyro/infer/mcmc/util.py (hunk 23 above)
Comment (neerajprad) [reply pyro/infer/mcmc/util.py]: done.

**Comment by neerajprad** [comment]
> Can you confirm that you have tests that would have caught the double-counting errors that plagued earlier versions of this PR?

Yes, I have added many tests that would catch such errors in `test_valid_models`. There were a couple of tests that would have caught this even earlier, had I stand guilty of having double counted myself while writing the expected values. :) 

> After this merges, I'd like to add some more interesting models as tests, e.g. HMM and GMM.

There is a small gmm test in `test_hmc.py`, but it'll be great to have HMM and other interesting models.

**Comment by neerajprad** [comment]
I think there is a dim error in the `logsumexp` implementation, which I'm looking into. Will update shortly.

**Comment by neerajprad** [comment]
Thanks a ton for the detailed review, @fritzo. I think I have addressed all comments, and fixed the `logsumexp` bug now.

**Comment by fritzo** [comment]
Nice work, @neerajprad!

**Comment by martinjankowiak** [comment]
yaaay!
this is awesome @neerajprad !
also, seems like this PR also benefited from some serious CR from @fritzo !

**Comment by eb8680** [comment]
Awesome!
